src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,162 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+
+import com.google.sps.data.Card;
+import com.google.sps.tool.WordSearch;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+
+/**
+ * Handles the study mode questions and responses by adjusting and sorting a card's based on their
+ * familiarity score and last time seen.
+ */
+@WebServlet("/study")
+public class StudyServlet extends HttpServlet {
+  public class QuizCard {
+    private String quizWord;
+    private String cardKey;
+    private ArrayList<String> possibleResponses = new ArrayList<>();
+    private String correctAnswer;
+
+    public QuizCard(Card card) {
+      this.quizWord = card.getTextTranslated();
+      this.cardKey = card.getCardKey();
+      this.correctAnswer = card.getRawText().toLowerCase();
+      this.possibleResponses = WordSearch.generateWordOptions(this.correctAnswer);
+    }
+  }
+
+  private int numOfCardsPerRound = 1;
+  private int maxNumOfRounds = 4;
+
+  // Returns QuizCard's that are sorted by familiarity score as json
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    List<Card> userCards = new ArrayList<>();
+
+    if (userService.isUserLoggedIn()) {
```

Comment:
Avoid indentations as much as possible. When you have this type of check, it's better to early return instead:

```java
if (!userService.isUserLoggedIn()) {
  // Handle logged out user...
  return;
}

// Handle logged in user...
```

Reply:
done.

Reply:
done.